torch/_inductor/decomposition.py

```diff
@@ -145,6 +145,113 @@ def addmm(input, mat1, mat2, *, beta=1, alpha=1):
     else:
         return NotImplemented  # go directly to lowering
 
+import time
+def bench(f, name=None, iters=100, warmup=5, display=True):
+    for _ in range(warmup):
+        f()
+    torch.cuda.synchronize()
+    begin = time.time()
+    for i in range(iters):
+        f()
+    torch.cuda.synchronize()
+    us_per_iter = (time.time()-begin)*1e6/iters
+    if name is None:
+        res = us_per_iter
+    else:
+        res= f"{name}: {us_per_iter}us"
+    if display:
+        print(res)
+    return res
+
+import torch.nn.functional as F
+rounding_factor = 8
+
+from torch.utils._mode_utils import no_dispatch
+def should_pad_bench(a, b, op):
+    with no_dispatch():
+        a = torch.randn_like(a)
+        b = torch.randn_like(b)
+        no_pad = bench(lambda: op(a, b), display=False)
+        def round_up(x):
+            return (x + rounding_factor - 1) // rounding_factor * rounding_factor
+        a_pad = a.new_empty([round_up(i) for i in a.shape])
+        b_pad = b.new_empty([round_up(i) for i in b.shape])
+        pad = bench(lambda: op(a_pad, b_pad), display=False)
+        # print(a.shape, b.shape)
+        # print(no_pad, pad)
+        return no_pad > pad + 100
+
+def should_pad_mm(a, b):
+    if all(x % rounding_factor == 0 for x in a.shape + b.shape):
+        return False
+    flops = a.shape[0] * a.shape[1] * b.shape[1]
+    bw = a.shape[0] * a.shape[1] + b.shape[0] * b.shape[1] + a.shape[0] * b.shape[1]
+    if flops <= bw * 64 or flops <= 1e9:
+        return False
```

Comment:
Do we need this heuristic if we also do autotuning?

Reply:
No, I started off writing this heuristic, but had some issues with actually getting it to work, so just wrote the autotuner to see what a perfect heuristic looks like.